Last-Update: 2014-12-25
Bug-Ceph: http://tracker.ceph.com/issues/10036
Author: Mykola Golub <mgolub@mirantis.com>
Description: show primary-affinity in OSD tree

--- a/src/osd/OSDMap.cc
+++ b/src/osd/OSDMap.cc
@@ -2349,19 +2349,23 @@
       *out << std::setprecision(4)
 	   << (exists(cur) ? get_weightf(cur) : 0)
 	   << std::setprecision(p)
 	   << "\t";
+      *out << std::setprecision(4)
+           << (exists(cur) ? get_primary_affinityf(cur) : 0)
+           << std::setprecision(p);
     }
     if (f) {
       f->dump_float("reweight", get_weightf(cur));
+      f->dump_float("primary_affinity", get_primary_affinityf(cur));
     }
   }
 }
 
 void OSDMap::print_tree(ostream *out, Formatter *f) const
 {
   if (out)
-    *out << "# id\tweight\ttype name\tup/down\treweight\n";
+    *out << "# id\tweight\ttype name\tup/down\treweight\tprimary-affinity\n";
   if (f)
     f->open_array_section("nodes");
   set<int> touched;
   set<int> roots;
