Last-Update: 2014-07-17
Forwarded: not-needed
Origin: upstream
Author: Dmitry Smirnov <onlyjob@member.fsf.org>
Description: fixes from "firefly" branch since 0.80.4 release

--- a/src/ceph-disk
+++ b/src/ceph-disk
@@ -2549,9 +2549,9 @@
 
 def main():
     args = parse_args()
 
-    loglevel = logging.INFO
+    loglevel = logging.WARNING
     if args.verbose:
         loglevel = logging.DEBUG
 
     logging.basicConfig(
--- a/src/init-ceph.in
+++ b/src/init-ceph.in
@@ -43,8 +43,12 @@
 . $LIBDIR/ceph_common.sh
 
 EXIT_STATUS=0
 
+# detect systemd
+SYSTEMD=0
+grep -qs systemd /proc/1/comm && SYSTEMD=1
+
 signal_daemon() {
     name=$1
     daemon=$2
     pidfile=$3
@@ -271,9 +275,13 @@
 
 	    [ -n "$wrap" ] && runmode="-f &" && runarg="-f"
 	    [ -n "$max_open_files" ] && files="ulimit -n $max_open_files;"
 
-	    cmd="$files $wrap $cmd --cluster $cluster $runmode"
+	    if [ $SYSTEMD -eq 1 ]; then
+		cmd="systemd-run -r bash -c '$files $cmd --cluster $cluster -f'"
+	    else
+		cmd="$files $wrap $cmd --cluster $cluster $runmode"
+	    fi
 
 	    if [ $dofsmount -eq 1 ] && [ -n "$fs_devs" ]; then
 		get_conf pre_mount "true" "pre mount command"
 		get_conf fs_type "" "osd mkfs type"
