Last-Update: 2014-10-02
Forwarded: not-needed
Origin: upstream
Author: Dmitry Smirnov <onlyjob@member.fsf.org>
Description: fixes from "firefly" branch since 0.80.6 release

--- a/src/common/config.cc
+++ b/src/common/config.cc
@@ -946,9 +946,9 @@
   return -ENOSYS;
 }
 
 static const char *CONF_METAVARIABLES[] =
-  { "cluster", "type", "name", "host", "num", "id", "pid" };
+  { "cluster", "type", "name", "host", "num", "id", "pid", "cctid" };
 static const int NUM_CONF_METAVARIABLES =
       (sizeof(CONF_METAVARIABLES) / sizeof(CONF_METAVARIABLES[0]));
 
 void md_config_t::expand_all_meta()
@@ -1058,8 +1058,10 @@
 	else if (var == "id")
 	  out += name.get_id().c_str();
 	else if (var == "pid")
 	  out += stringify(getpid());
+	else if (var == "cctid")
+	  out += stringify((unsigned long long)this);
 	else
 	  assert(0); // unreachable
 	expanded = true;
       }
