Last-Update: 2014-10-31
Forwarded: no
Author: Dmitry Smirnov <onlyjob@member.fsf.org>
Description: disable tests that depend on network...

--- a/src/test/Makefile.am
+++ b/src/test/Makefile.am
@@ -230,8 +230,9 @@
 
 
 ## Unit tests
 
+if WITH_CLUSTER_TESTS
 check_SCRIPTS += \
 	unittest_bufferlist.sh \
 	test/encoding/check-generated.sh \
 	test/mon/osd-pool-create.sh \
@@ -243,8 +244,9 @@
 	test/osd/osd-bench.sh \
 	test/ceph-disk.sh \
 	test/mon/mon-handle-forward.sh \
 	test/vstart_wrapped_tests.sh
+endif
 
 EXTRA_DIST += \
 	$(srcdir)/test/mon/mon-test-helpers.sh \
 	$(srcdir)/test/osd/osd-test-helpers.sh \
--- a/src/test/erasure-code/Makefile.am
+++ b/src/test/erasure-code/Makefile.am
@@ -1,6 +1,8 @@
+if WITH_CLUSTER_TESTS
 check_SCRIPTS += \
 	test/erasure-code/test-erasure-code.sh
+endif
 
 ceph_erasure_code_benchmark_SOURCES = \
 	erasure-code/ErasureCode.cc \
 	test/erasure-code/ceph_erasure_code_benchmark.cc
--- a/configure.ac
+++ b/configure.ac
@@ -951,8 +951,15 @@
 AC_LANG_POP([C++])
 
 AM_CONDITIONAL(WITH_BUILD_TESTS, test "$WITH_BUILD_TESTS" = "1")
 
+AC_ARG_WITH([cluster-tests],
+    [AS_HELP_STRING([--with-cluster-tests], [enables cluster tests])],
+    [],
+    [with_cluster_tests=no]
+)
+AM_CONDITIONAL(WITH_CLUSTER_TESTS, [ test "$with_cluster_tests" = "yes" ])
+
 AM_PATH_PYTHON([2.4],
 	[], [AC_MSG_FAILURE([Failed to find Python 2.4 or newer])])
 
 AC_CONFIG_HEADERS([src/acconfig.h])
