Description: Ensure systemd used >= vivid
Author: James Page <james.page@ubuntu.com>
Forwarded: no

--- a/src/ceph-disk
+++ b/src/ceph-disk
@@ -2128,7 +2128,10 @@ def activate(
             else:
                 (distro, release, codename) = platform.dist()
                 if distro == 'Ubuntu':
-                    init = 'upstart'
+                    if codename >= 'vivid':
+                        init = 'systemd'
+                    else:
+                        init = 'upstart'
                 else:
                     init = 'sysvinit'
 
